lib/Sema/ConstraintSystem.h

```diff
@@ -1019,6 +1039,11 @@ class Solution {
   /// to make the solution work.
   llvm::SmallVector<ConstraintFix *, 4> Fixes;
 
+  /// For locators associated with call expressions, the trailing closure
+  /// matching rule that was applied.
+  llvm::SmallMapVector<ConstraintLocator*, TrailingClosureMatching, 4>
+    trailingClosureMatchingChoices;
```

Comment:
I think that this could be avoided if the matching of the argument in `Backward` mode always resulted in the "warning" fix being recorded in the constraint system. Such fixes do not interfere with solution application and at the same time would record all of the information necessary to produce a warning e.g. which parameter did get bound to the trailing closure. This would also help remove most of the new logic from CSApply which tried to determine what happened with trailing closure.